openstack/standalone-cinder/mapper-iscsi.go

```diff
@@ -0,0 +1,107 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"k8s.io/api/core/v1"
+	"github.com/golang/glog"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+
+const ISCSI_TYPE = "iscsi"
+const INITIATOR_NAME = "iqn.1994-05.com.redhat:a13fc3d1cc22"
+
+
+type iscsiMapper struct {
+	volumeMapper
+}
+
+
+func getChapSecretName(ctx provisionCtx) string {
+	if ctx.connection.Data.AuthMethod == "CHAP" {
+		return ctx.options.PVName + "-secret"
+	} else {
+		return ""
+	}
+}
+
+func (m *iscsiMapper) BuildPVSource(ctx provisionCtx) (*v1.PersistentVolumeSource, error) {
+	ret := &v1.PersistentVolumeSource{
+		ISCSI: &v1.ISCSIVolumeSource{
+			// TODO: Need some way to specify the initiator name
+			TargetPortal: ctx.connection.Data.TargetPortal,
+			IQN: ctx.connection.Data.TargetIqn,
+			Lun: ctx.connection.Data.TargetLun,
+			SessionCHAPAuth: false,
+		},
+	}
+	secretName := getChapSecretName(ctx)
+	if secretName != "" {
+		ret.ISCSI.SessionCHAPAuth = true
+		ret.ISCSI.SecretRef = &v1.LocalObjectReference{
+			Name: secretName,
+		}
+	}
+	return ret, nil
+}
+
+func (m *iscsiMapper) AuthSetup(ctx provisionCtx) error {
```

Comment:
iSCSI CHAP is only in 1.7+

Reply:
How should I handle the kubernetes version dependencies for this?

Reply:
server git version should help, see
https://github.com/kubernetes-incubator/external-storage/blob/master/lib/controller/controller.go#L370